server/swap/swap.go

```diff
@@ -601,172 +433,125 @@ func (s *Swapper) restoreState(state *State, allowPartial bool) error {
 		return nil
 	}
 
-	s.matches = make(map[order.MatchID]*matchTracker, len(state.MatchTrackers))
+	s.matches = make(map[order.MatchID]*matchTracker, len(swapData))
 	s.userMatches = make(map[account.AccountID]map[order.MatchID]*matchTracker)
-	for mid, mtd := range state.MatchTrackers {
-		// Check and skip matches for missing assets.
-		makerSwapAsset := mtd.MakerStatus.SwapAsset
-		makerRedeemAsset := mtd.MakerStatus.RedeemAsset
-		if missingAssets[makerSwapAsset] {
-			log.Infof("Skipping match %v with missing asset %d", mid, makerSwapAsset)
+	for _, sd := range swapData {
+		if missingAssets[sd.Base] {
+			log.Warnf("Dropping match %v with no backend available for base asset %d", sd.ID, sd.Base)
 			continue
 		}
-		if missingAssets[makerRedeemAsset] {
-			log.Infof("Skipping match %v with missing asset %d", mid, makerRedeemAsset)
+		if missingAssets[sd.Quote] {
+			log.Warnf("Dropping match %v with no backend available for quote asset %d", sd.ID, sd.Quote)
 			continue
 		}
-
-		mt := &matchTracker{
-			Match:     mtd.Match,
-			time:      encode.UnixTimeMilli(mtd.Time),
-			matchTime: mtd.Match.Epoch.End(),
-			makerStatus: &swapStatus{
-				swapAsset:   makerSwapAsset,
-				redeemAsset: makerRedeemAsset,
-			},
-			takerStatus: &swapStatus{
-				swapAsset:   makerRedeemAsset, // mtd.TakerStatus.SwapAsset
-				redeemAsset: makerSwapAsset,   // mtd.TakerStatus.RedeemAsset
-			},
+		// Load the maker's order.LimitOrder and taker's order.Order.
+		taker, _, err := s.storage.Order(sd.MatchData.Taker, sd.Base, sd.Quote)
+		if err != nil {
+			log.Errorf("Failed to load taker order: %v", err)
+			continue
```

Comment:
Noting that this is a departure from normal because the `Order` we'll be tracking in `Swapper` is not backed by the same underlying being tracked in `Market`. Is that okay?  I think the only thing we'd need to worry about is if we were somehow relying on the `Filled/Remaining` quantities to remain accurate, and a quick scan seems to say that we are not.

Reply:
Right, good catch and it's part of why it's better that market is back in charge of order status fully again.  I will certainly make a note about the caveat you identified and have a scan myself.

Reply:
Turned out that `(*Market).SwapDone` was calling `Remaining()`.  Fortunately since it's in Market, we don't have to infer based on type/force/fill because we can just to `Book.HaveOrder()`, so that was an easy fix.

Probably safest would be not to have and `order.Order` as an argument, and instead pass a whole bunch of separate data (ID, type, force, base, quote, user), but we'd need to hack together an Order anyway to use the coin unlocking and storage functions and I think that's even more error prone.  We're returning to the idea of a `FilledOrder` that adds the annoying Filled/Remaining stuff so that the order.Order type can be immutable..  Maybe in the future.  It's a big change to do right.